File: frontend/src/types/apiKey.ts
/** Mirrors app/core/scopes.py::ApiKeyScope. */
export const API_KEY_SCOPES = [
  "organizations:read",
  "workspaces:read",
  "workspaces:write",
  "members:read",
  "work_items:read",
  "work_items:write",
  "audit_logs:read",
  "files:read",
  "files:write",
  "webhooks:read",
  "webhooks:write",
  "webhooks:admin",
  "billing:read",
] as const;

export type ApiKeyScope = (typeof API_KEY_SCOPES)[number];

export interface ApiKeyRead {
  readonly id: string;
  readonly organization_id: string;
  readonly user_id: string | null;
  readonly name: string;
  readonly scopes: readonly string[];
  readonly last_used_at: string | null;
  readonly expires_at: string | null;
  readonly deactivated_at: string | null;
  readonly deactivated_reason: string | null;
  readonly previous_secret_expires_at: string | null;
  readonly created_at: string;
  readonly updated_at: string;
}

export interface ApiKeyResponse {
  readonly api_key: ApiKeyRead;
  readonly token: string;
}

export interface ApiKeyCreateRequest {
  readonly name: string;
  readonly scopes: readonly ApiKeyScope[];
  readonly expires_at?: string | null;
}

File: backend/app/schemas/api_key.py
"""API Key validation and serialization contracts (ARCH-08 §9.6)."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from app.core.scopes import ApiKeyScope


class ApiKeyCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=120)
    scopes: list[ApiKeyScope] = Field(..., min_items=1)
    expires_at: Optional[datetime] = None


class ApiKeyRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    organization_id: uuid.UUID
    user_id: Optional[uuid.UUID] = None
    name: str
    scopes: list[str]
    last_used_at: Optional[datetime] = None
    expires_at: Optional[datetime] = None
    deactivated_at: Optional[datetime] = None
    deactivated_reason: Optional[str] = None
    previous_secret_expires_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class ApiKeyResponse(BaseModel):
    """Returned ONCE upon issuance. Contains the full token string."""
    api_key: ApiKeyRead
    token: str = Field(description="Full token string. Shown ONCE. Never stored or logged.")


class ApiKeyRotateRequest(BaseModel):
    force: bool = Field(default=False, description="Force rotation even if a dual-secret overlap is currently active.")
